cmd/ocsp-updater/main.go

```diff
@@ -107,97 +203,105 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
 	}
+	updater.stats.Inc("OCSP.StoredResponses", int64(len(responses)), 1.0)
 
-	updater.log.Info(fmt.Sprintf("OCSP %s: OK", serial))
-	updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
-	updater.stats.TimingDuration("OCSP.Updates.UpdateLatency", time.Since(innerStart), 1.0)
-	return nil
+	return
 }
 
-// findStaleResponses opens a transaction and processes up to responseLimit
-// responses in a single batch. The responseLimit should be relatively small,
-// so as to limit the chance of the transaction failing due to concurrent
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
+// oldOCSPResponsesTick looks for certificates with stale OCSP responses and
+// generates/stores new ones
+func (updater *OCSPUpdater) oldOCSPResponsesTick(batchSize int) {
+	now := time.Now()
+	statuses, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry), batchSize)
+	if err != nil {
+		updater.stats.Inc("OCSP.Errors.FindStaleResponses", 1, 1.0)
+		updater.log.AuditErr(fmt.Errorf("Failed to find stale OCSP responses: %s", err))
+		return
+	}
 
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
-			}
-		}
+	updater.generateOCSPResponses(statuses)
+}
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
-	}
+type looper struct {
+	clk       clock.Clock
+	stats     statsd.Statter
+	batchSize int
+	tickDur   time.Duration
+	tickFunc  func(int)
+	name      string
+}
 
-	return err
+func (l *looper) loop() {
+	for {
+		tickStart := l.clk.Now()
+		l.tickFunc(l.batchSize)
+		l.stats.TimingDuration(fmt.Sprintf("OCSP.%s.TickDuration", l.name), time.Since(tickStart), 1.0)
+		l.stats.Inc(fmt.Sprintf("OCSP.%s.Ticks", l.name), 1, 1.0)
+		tickEnd := tickStart.Add(time.Since(tickStart))
+		expectedTickEnd := tickStart.Add(l.tickDur)
+		shortTickEnd := tickStart.Add(l.tickDur / -2)
+		if tickEnd.After(expectedTickEnd) {
+			l.stats.Inc(fmt.Sprintf("OCSP.%s.LongTicks", l.name), 1, 1.0)
+		} else if tickEnd.Before(shortTickEnd) {
```

Comment:
per offline discussion, delete shortTickEnd (because it's not an error case), and just always sleep til expectedTickEnd (using the clock code that fixes the fake in the negative sleep case).